File: load_MEP.m
% input
%   subj: subject index 
%   iidx: intensity index 
%
% output
%   y: target signal [n x 300] if tcrop=[20,50]ms
%   t: [1 x 300]
%   yall: [n x 300 x 15 trials]
%   mep: all MEPs [N intensities x 10000 timepoints x 15 trials]
%   intensities: all intensities [1 x N]
%   times: [1 x 10000], dt=0.1ms
%
% example
%   subj=1; load_MEP(subj);
%   subj=1; iidx=[5,6];  load_MEP(subj,iidx);
function [y,t,mep,intensities,times,yall] = load_MEP(subj,iidx,tcrop,plotOn)
    if nargin < 3
        tcrop = [20 50]; %ms  
    end
    if nargin < 4
        plotOn=1;
    end
    
    TSTIM=100; % TMS stimulus was at t=100ms
    root=fileparts(mfilename("fullpath"));
    tmp=load(fullfile(root,'data_Oxford_MEP',...
             sprintf('S%d_Magstim_data',subj),...
             sprintf('S%d.mat',subj))); 
    mep=tmp.mep;
    intensities=tmp.intensities;
    times=tmp.t-TSTIM;
    if nargin < 2
        iidx=1:length(intensities); % all intensities
    end
    mep=mep(iidx,:,:);
    intensities=intensities(iidx);

    % crop   
    tidx=find(times>=tcrop(1) & times<tcrop(2));
    t = times(tidx); 
    yall=mep(iidx,tidx,:);
    y = mean(yall,3);

    % remove baseline [-20 -10]msec % especially for subject 6
    baseline=mean(mean(mep(iidx,times>=-20 & times<-10,:),3),2);
    y=y-repmat(baseline,[1,size(y,2)]);   
    mep=mep-repmat(baseline,[1,size(mep,2),size(mep,3)]);

    if plotOn
        figure % all
        for i=1:length(intensities)
          subplot(2,5,i)
          plot(times,squeeze(mep(i,:,:)),'c');hold on;grid on;
          plot(times,mean(squeeze(mep(i,:,:)),2),'k','linewidth',1.5);
          xlim([20 50])
          ylim([-2 5])
          title(sprintf('%g%% MSO',intensities(i)));
          xlabel('msec')
        end

        figure % selected
        plot(t,y,'linewidth',1.5);grid on;
        title(sprintf('Subject %d',subj))
        xlabel('Time (msec)');ylabel('Amplitude (mV)')
  
        legend_labels = strcat(string(intensities), '%MSO');
        legend(legend_labels)
    end
end

File: load_MEP.py
import os
import h5py
import numpy as np
import numpy as np
import matplotlib.pyplot as plt
from load_h5 import load_h5_to_dict


def load_MEP(subj, iidx=None, tcrop=[20, 50], plotOn=1):
    TSTIM = 100  # TMS stimulus was at t=100 ms

    # ----- load data -----
    root = os.getcwd()
    mat_path = os.path.join(
        root,
        "data_Oxford_MEP",
        f"S{subj}_Magstim_data",
        f"S{subj}.h5"
    )
    
    with h5py.File(mat_path, 'r') as f:
        tmp = load_h5_to_dict(f)

    mep = tmp["mep"]                 # shape: [N x time x trials]
    intensities = tmp["intensities"].flatten()
    times = tmp["t"].flatten() - TSTIM

    # ----- select intensities -----
    if iidx is None:
        iidx = np.arange(len(intensities))

    mep = mep[iidx, :, :]
    intensities = intensities[iidx]

    # ----- crop -----
    tidx = np.where((times >= tcrop[0]) & (times < tcrop[1]))[0]
    t = times[tidx]

    yall = mep[:, tidx, :]
    y = np.mean(yall, axis=2)

    # ----- remove baseline [-20, -10] ms -----
    baseline_idx = np.where((times >= -20) & (times < -10))[0]

    baseline = np.mean(
        np.mean(mep[:, baseline_idx, :], axis=2),
        axis=1
    )  # shape: [n]

    # subtract baseline
    y = y - baseline[:, np.newaxis]
    mep = mep - baseline[:, np.newaxis, np.newaxis]

    # ----- plotting -----
    if plotOn:

        # figure 1: all trials
        plt.figure()
        for i in range(len(intensities)):
            plt.subplot(2, 5, i + 1)

            plt.plot(times, mep[i, :, :], 'c', linewidth=0.8)
            plt.grid(True)

            mean_trace = np.mean(mep[i, :, :], axis=1)
            plt.plot(times, mean_trace, 'k', linewidth=1.5)

            plt.xlim([20, 50])
            plt.ylim([-2, 5])

            plt.title(f"{intensities[i]}% MSO")
            plt.xlabel("msec")

        # figure 2: averaged signals
        plt.figure()
        for i in range(len(intensities)):
            plt.plot(t, y[i, :], linewidth=1.5)

        plt.grid(True)
        plt.title(f"Subject {subj}")
        plt.xlabel("Time (msec)")
        plt.ylabel("Amplitude (mV)")
        plt.xlim([20, 50])

        legend_labels = [f"{val}%MSO" for val in intensities]
        plt.legend(legend_labels)

        plt.show()

    return y, t, mep, intensities, times, yall